coderd/notifications/manager.go

```diff
@@ -0,0 +1,363 @@
+// succeeded or failed every CODER_NOTIFICATIONS_STORE_SYNC_INTERVAL seconds.
+// These buffers are limited in size, and naturally introduce some backpressure; if there are hundreds of messages to be
+// sent but they start failing too quickly, the buffers (receive channels) will fill up and block senders, which will
+// slow down the dispatch rate.
+//
+// NOTE: The above backpressure mechanism only works within the same process, which may not be true forever, such as if
+// we split notifiers out into separate targets for greater processing throughput; in this case we will need an
+// alternative mechanism for handling backpressure.
+type Manager struct {
+	cfg codersdk.NotificationsConfig
+
+	store Store
+	log   slog.Logger
+
+	notifier *notifier
+	handlers map[database.NotificationMethod]Handler
+
+	runOnce  sync.Once
+	stopOnce sync.Once
+	stop     chan any
+	done     chan any
+}
+
+// NewManager instantiates a new Manager instance which coordinates notification enqueuing and delivery.
+//
+// helpers is a map of template helpers which are used to customize notification messages to use global settings like
+// access URL etc.
+func NewManager(cfg codersdk.NotificationsConfig, store Store, log slog.Logger) (*Manager, error) {
+	// If dispatch timeout exceeds lease period, it is possible that messages can be delivered in duplicate because the
+	// lease can expire before the notifier gives up on the dispatch, which results in the message becoming eligible for
+	// being re-acquired.
+	if cfg.DispatchTimeout.Value() >= cfg.LeasePeriod.Value() {
+		return nil, ErrInvalidDispatchTimeout
+	}
+
+	return &Manager{
+		log:   log,
+		cfg:   cfg,
+		store: store,
+
+		stop: make(chan any),
+		done: make(chan any),
+
+		handlers: defaultHandlers(cfg, log),
+	}, nil
+}
+
+// defaultHandlers builds a set of known handlers; panics if any error occurs as these handlers should be valid at compile time.
+func defaultHandlers(cfg codersdk.NotificationsConfig, log slog.Logger) map[database.NotificationMethod]Handler {
+	return map[database.NotificationMethod]Handler{
+		database.NotificationMethodSmtp:    dispatch.NewSMTPHandler(cfg.SMTP, log.Named("dispatcher.smtp")),
+		database.NotificationMethodWebhook: dispatch.NewWebhookHandler(cfg.Webhook, log.Named("dispatcher.webhook")),
+	}
+}
+
+// WithHandlers allows for tests to inject their own handlers to verify functionality.
+func (m *Manager) WithHandlers(reg map[database.NotificationMethod]Handler) {
+	m.handlers = reg
+}
+
+// Run initiates the control loop in the background, which spawns a given number of notifier goroutines.
+// Manager requires system-level permissions to interact with the store.
+// Run is only intended to be run once.
+func (m *Manager) Run(ctx context.Context) {
+	m.log.Info(ctx, "started")
+
+	m.runOnce.Do(func() {
+		// Closes when Stop() is called or context is canceled.
+		go func() {
+			err := m.loop(ctx)
```

Comment:
If `loop` accidentally dies/panics, the coderd will still be up and running, right? Should we refactor this into a supervisor pattern that will boot another `loop` thread or technically it shouldn't be possible?

Reply:
If it panics, it'll kill `coderd`; this isn't unique to this component though.
We probably don't want to auto-recover since `panic`s are normally for a good reason and can lead to a crashloop.